src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/CompletionHandler.cs

```diff
@@ -152,6 +163,58 @@ internal class CompletionHandler : IRequestHandler<CompletionParams, SumType<Com
             return result;
         }
 
+        private bool IsSimpleImplicitExpression(LSPDocumentSnapshot documentSnapshot, CompletionParams request, LanguageServerKind serverKind)
+        {
+            if (serverKind != LanguageServerKind.CSharp)
+            {
+                return false;
+            }
+
+            if (string.Equals(request.Context.TriggerCharacter, "@", StringComparison.Ordinal))
+            {
+                // Completion was triggered with `@` this is always a simple implicit expression
+                return true;
+            }
+
+            var snapshot = documentSnapshot.Snapshot;
+            var navigator = _textStructureNavigator.GetTextStructureNavigator(snapshot.TextBuffer);
+            var line = snapshot.GetLineFromLineNumber(request.Position.Line);
+            var absoluteIndex = line.Start + request.Position.Character;
+            if (absoluteIndex >= snapshot.Length)
+            {
+                Debug.Fail("This should never happen when resolving C# polyfills given we're operating on snapshots.");
+                return false;
+            }
+
+            // Lets walk backwards to the character that caused completion (if one triggered it) to ensure that the "GetExtentOfWord" returns
+            // the word we care about and not whitespace following it. For instance:
+            //
+            //      @Date|\r\n
+            //
+            // Will return the \r\n as the "word" which is incorrect; however, this is actually fixed in newer VS CoreEditor builds but is behind
+            // the "Use word pattern in LSP completion" preview feature. Once this preview feature flag is the default we can remove this -1.
+            var completionCharacterIndex = Math.Max(0, absoluteIndex - 1);
+            var completionSnapshotPoint = new SnapshotPoint(documentSnapshot.Snapshot, completionCharacterIndex);
+            var wordExtent = navigator.GetExtentOfWord(completionSnapshotPoint);
+
+            if (!wordExtent.IsSignificant)
+            {
+                // Word is only whitespace, definitely not an implicit expresison
+                return false;
+            }
+
+            // We need to look at the item before the word because `@` at the beginning of a word is not encapsulated in that word.
```

Comment:
Is `@` not encapsulated in the word in both cases with and without the feature flag?

Reply:
Not in implicit expression scenarios. Keep in mind we opted in to using the default word pattern which doesn't include `@` 😄